core-bundle/src/EventListener/BackendMenuListener.php

```diff
@@ -12,82 +12,52 @@
 
 namespace Contao\CoreBundle\EventListener;
 
-use Contao\BackendUser;
+use Contao\Backend;
 use Contao\CoreBundle\Event\MenuEvent;
-use Knp\Menu\FactoryInterface;
-use Knp\Menu\ItemInterface;
+use Contao\CoreBundle\Framework\ContaoFramework;
 use Symfony\Component\Security\Core\Security;
 
-class BackendMenuListener
+final class BackendMenuListener
 {
     /**
      * @var Security
      */
     private $security;
 
-    public function __construct(Security $security)
+    /**
+     * @var ContaoFramework
+     */
+    private $framework;
+
+    public function __construct(Security $security, ContaoFramework $framework)
     {
         $this->security = $security;
+        $this->framework = $framework;
     }
 
     /**
-     * Adds the back end user navigation.
+     * Adds the number of system error messages to the alerts link.
      */
     public function onBuild(MenuEvent $event): void
     {
-        $user = $this->security->getUser();
-
-        if (!$user instanceof BackendUser) {
+        if (!$this->security->isGranted('ROLE_ADMIN')) {
             return;
         }
 
-        $factory = $event->getFactory();
-        $tree = $event->getTree();
-        $modules = $user->navigation();
-
-        foreach ($modules as $categoryName => $categoryData) {
-            $categoryNode = $tree->getChild($categoryName);
-
-            if (!$categoryNode) {
-                $categoryNode = $this->createNode($factory, $categoryName, $categoryData);
-
-                if (isset($categoryData['class']) && preg_match('/\bnode-collapsed\b/', $categoryData['class'])) {
-                    $categoryNode->setDisplayChildren(false);
-                }
+        $item = $event->getTree()->getChild('alerts');
 
-                $tree->addChild($categoryNode);
-            }
-
-            // Create the child nodes
-            foreach ($categoryData['modules'] as $moduleName => $moduleData) {
-                $moduleNode = $this->createNode($factory, $moduleName, $moduleData);
-                $moduleNode->setCurrent((bool) $moduleData['isActive']);
-
-                $categoryNode->addChild($moduleNode);
-            }
+        if (null === $item) {
+            return;
         }
-    }
 
-    private function createNode(FactoryInterface $factory, string $name, array $attributes): ItemInterface
-    {
-        $classes = [];
+        /** @var Backend $adapter */
+        $adapter = $this->framework->getAdapter(Backend::class);
+        $count = substr_count($adapter->getSystemMessages(), 'class="tl_error');
```

Comment:
Previously, there was a separation between the system message count and the error count. Can you explain why that's no longer the case?

Reply:
Because `$systemMessagesCount` is not used anywhere (not even in Contao 4.8). Seemed like a left-over, therefore I have not added it in the new class.

Reply:
And we're not showing non-error messages anymore either?

Reply:
We are showing them in the popup window, but the counter of the alerts button only shows error messages. This is the current behaviour in Contao 4.4+.

https://github.com/contao/contao/blob/4.4/core-bundle/src/Resources/contao/templates/backend/be_main.html5

Reply:
But if there is no count on error messages, we do not generate the menu item at all? I think that was also not the case previously, because we always showed the popup link?

Reply:
Of course we still show the menu item; it is always added in the `BackendMenuBuilder`. The listener just enhances the menu by adding the system error messages count.

Reply:
Can you please explain the code to me? If `$count` is smaller than `1`, the method returns. Otherwise it does `$factory->createItem()`. Isnt' that `$factory->createItem()` related to adding the notification item at all?

PS: please don't mark _my_ questions as resolved, if _you_ answer them. I think only the original author should do that, except if e.g. you commit a change that solves the question.

Reply:
> If `$count` is smaller than `1`, the method returns. Otherwise it does `$factory->createItem()`

If `$count` is smaller than `1`, the method returns. Otherwise it does `$item->setLabel()` and not `$factory->createItem()`! You should review more carefully.

Reply:
oh well… the diff showed this as the old code, must have gotten me confused 😂